pkg/engine/constraints.go

```diff
@@ -0,0 +1,127 @@
+package engine
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/klothoplatform/klotho/pkg/core"
+	"github.com/klothoplatform/klotho/pkg/graph"
+	"gopkg.in/yaml.v3"
+)
+
+type (
+	Constraint interface {
+		Level() ConstraintLevel
+		IsSatisfied(dag *core.ResourceGraph) bool
+		CanApply() bool
```

Comment:
What does `CanApply` mean? Does it mean "this is potentially relevant", or "this is potentially satisfiable"?

Looking at where it's used below, it seems like it's "this is potentially relevant". If so, do we really need this method? Why not just let `Apply(dag)` determine that? If the constraint isn't applicable, `Apply` can short-circuit return a no-error with no modification.

Similarly, I don't understand what `IsSatisfied` is for.

Reply:
so we could also do that, this part wont make sense yet as its me thinking about the engine and not the api portion.

But my thought was

IsSatisfied is just a method which checks to see if the dag satisfies the constraint
CanApply is a method that sees if the constraint is a type of constraint that we can satisfy by making changes to the dag. This might not be necessary since we should jsut be able to run Apply, but not important yet, its all implementation details behind the scene. Was just trying to start thinking about interfaces and stuff for our go layer

Reply:
On that note, could you add godoc comments explaining what each function on the interface does and how we (initially) intend to use it?

Reply:
yeah ill definitely add some and brush that up to make it clear. This pr when i move it out of draft will also just purely be constructs and parsing them from a yaml file so a lot of it will be clearer